File: matlab/BezierFitDemo_debug.m
function BezierFitDemo_debug
% BezierFitDemo_debug.m - Versão com prints de debug
% Baseado no código original BezierFitDemo.m

clear;
demo = 1;

switch demo
    case 1
        C =[0     0;
            1     2;
            3     3;
            4     2];
        
        Q = cubicBezierToPolyline(C, 65);
        n = 3;  % Starting number of knot points
end

fprintf('=== MATLAB DEBUG ===\n');
fprintf('Demo: %d, n: %d\n', demo, n);
fprintf('Q shape: %dx%d\n', size(Q,1), size(Q,2));
fprintf('Q first 5 points:\n');
disp(Q(1:5,:));

% Now try to do a piecewise cubic Bézier fit to Q starting with n knot points
Qt = Q';
fprintf('Qt shape: %dx%d\n', size(Qt,1), size(Qt,2));

[IG, k] = iguess0_debug(Qt, n);

fprintf('=== AFTER IGUESS0 ===\n');
fprintf('IG length: %d\n', length(IG));
fprintf('k: ');
disp(k);
fprintf('IG: ');
disp(IG);

% Improve the fit: Segmentally Optimum Only Curve (SOO)
SOC = segop_debug(k, Qt, IG);

fprintf('=== AFTER SEGOP ===\n');
fprintf('SOC length: %d\n', length(SOC));
fprintf('SOC: ');
disp(SOC);

% Improve it again: Segmentally then Globally Optimized Curve (SGO)
GOC = globop_debug(SOC, Qt, 0, k);

fprintf('=== AFTER GLOBOP ===\n');
fprintf('GOC length: %d\n', length(GOC));
fprintf('GOC: ');
disp(GOC);

fprintf('=== MATLAB DEBUG END ===\n');
end

File: scripts/BezierFitDemo_debug.py
import numpy as np
import matplotlib
matplotlib.use('Agg')  # Use non-interactive backend
import matplotlib.pyplot as plt

from iguess0 import iguess0
from segop import segop  
from globop import globop
from cubicBezierToPolyline import cubicBezierToPolyline

def BezierFitDemo_debug():
    print("=== PYTHON DEBUG ===")
    
    demo = 1
    k = None

    if demo == 1:
        C = np.array([[0, 0],
                      [1, 2],
                      [3, 3],
                      [4, 2]])
        Q = cubicBezierToPolyline(C, 65)
        n = 3  # Starting number of knot points
        
    print(f"Demo: {demo}, n: {n}")
    print(f"Q shape: {Q.shape[0]}x{Q.shape[1]}")
    print("Q first 5 points:")
    print(Q[:5,:])

    # Now try to do a piecewise cubic Bézier fit to Q starting with n knot points
    Qt = Q.T
    print(f"Qt shape: {Qt.shape[0]}x{Qt.shape[1]}")

    IG, k, dpkpc = iguess0(Qt, n, k)

    print("=== AFTER IGUESS0 ===")
    print(f"IG length: {len(IG)}")
    print("k:", k)
    print("IG:", IG)

    # Improve the fit: Segmentally Optimum Only Curve (SOO)
    SOC = segop(k, Qt, IG)

    print("=== AFTER SEGOP ===")
    print(f"SOC length: {len(SOC)}")
    print("SOC:", SOC)

    # Improve it again: Segmentally then Globally Optimized Curve (SGO)
    GOC = globop(SOC, Qt, 0, k, dpkpc)

    print("=== AFTER GLOBOP ===")
    print(f"GOC length: {len(GOC)}")
    print("GOC:", GOC)

    print("=== PYTHON DEBUG END ===")

if __name__ == '__main__':
    BezierFitDemo_debug()
